feat(curriculum): create javascript fundamentals quiz (#56565)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/quiz-javascript-fundamentals/66edcd875b0d91de1fbbb492.md

#### --text--

> Default parameters in JavaScript refer to:

Reviewer: ```suggestion
What are default parameters?
```

Reviewer: This one has not been applied.

#### --distractors--

Parameters that should always be present in a function

---

Parameters that are passed in an array

---

Parameters that are destructured from an object

#### --answer--

Parameters that are optional and have a default value when not provided

### --question--

#### --text--

What is the correct syntax for a default parameter in JavaScript?

#### --distractors--

> function myFunction(param1 == defaultValue)

Reviewer: ```suggestion
`function myFunction(param1 == defaultValue)`
```

---

> function myFunction(param1: defaultValue)

Reviewer: ```suggestion
`function myFunction(param1: defaultValue)`
```

---

> function myFunction(param1 || defaultValue)

Reviewer: ```suggestion
`function myFunction(param1 || defaultValue)`
```

#### --answer--

> function myFunction(param1 = defaultValue)

Reviewer: ```suggestion
`function myFunction(param1 = defaultValue)`
```

### --question--

#### --text--

> What does the 'rest parameter' syntax allow us to do in JavaScript?

Reviewer: ```suggestion
What does the rest parameter syntax allow you to do in JavaScript?
```

#### --distractors--

Pass parameters by reference

---

Provide a set default parameters

---

Destructure an object into parameters

#### --answer--

Capture all arguments into one array

### --question--

#### --text--

Which of the following is the correct syntax for the rest parameter in JavaScript?

#### --distractors--

> function myFunction(restParams)

Reviewer: ```suggestion
`function myFunction(restParams)`
```

---

> function myFunction(restParams...)

Reviewer: ```suggestion
`function myFunction(restParams...)`
```

---

> function myFunction([...restParams])

Reviewer: ```suggestion
`function myFunction([...restParams])`
```

#### --answer--

> function myFunction(...restParams)

Reviewer: ```suggestion
`function myFunction(...restParams)`
```

### --question--

#### --text--

> What does destructuring allows us to do in JavaScript?

Reviewer: ```suggestion
What does destructuring allow you to do in JavaScript?
```

#### --distractors--

Build complex objects

---

> Serialize our objects

Reviewer: ```suggestion
Serialize objects
```

---

Merge arrays or objects

#### --answer--

Extract properties from objects and elements from arrays

### --question--

#### --text--

Which of the following is a valid function name in JavaScript?

#### --distractors--

> 123function

Reviewer: ```suggestion
`123function`
```

---

> function-name

Reviewer: ```suggestion
`function-name`
```

---

> function

Reviewer: ```suggestion
`function`
```

#### --answer--

> functionName

Reviewer: ```suggestion
`functionName`
```

### --question--

#### --text--

Which of these is NOT a good practice for naming variables and functions in JavaScript?

#### --distractors--

Use camelCase notation

---

Start variable names with a letter, $ or _

---

Using meaningful names

#### --answer--

Using reserved keywords as variable names

### --question--

#### --text--

What is the primary purpose of a linter in JavaScript development?

#### --distractors--

To format code automatically

---

To optimize code for performance

---

To create documentation for code

#### --answer--

To detect potential errors and style issues in code

### --question--

#### --text--

Which of the following is NOT a linter for JavaScript?

#### --distractors--

ESLint

---

standardJS

---

quick-lint-js

#### --answer--

Babel

### --question--

#### --text--

> Which of these describes accessibility

Reviewer: ```suggestion
Which of these describes accessibility?
```

#### --distractors--

Making sure code is accessible to other developers

---

Writing secure code

---

Controlling who can access certain data

#### --answer--

> Making web interfaces usable by all, including people with disabilities

Reviewer 2: I think this question will be better if it dealt with Accessibility in a Javascript context rather than a general one. 

https://developer.mozilla.org/en-US/curriculum/core/accessibility/#7.3_accessible_javascript

### --question--

#### --text--

> Why is using the 'var' keyword considered bad practice in modern JavaScript?

Reviewer: ```suggestion
Why is using the `var` keyword considered bad practice in modern JavaScript?
```

#### --distractors--

> It's slower than let and const

Reviewer: ```suggestion
It's slower than `let` and `const`
```

---

It causes a syntax error

---

It doesn't allow type checking

#### --answer--

It has function scope instead of block scope, which can lead to unexpected behaviors

### --question--

#### --text--

In JavaScript, how is memory for variables typically allocated?

#### --distractors--

Manually by the developer

---

By the browser

---

By the server

#### --answer--

> Automatically by the garbage collector

Reviewer 3: Not sure if this is right. I always thought garbage collector is not responsible for allocating as well.

Reviewer: Correct. The garbage collector reclaims memory.

> Automatically by the garbage collector

Reviewer: ```suggestion
Automatically by the runtime engine
```

### --question--

#### --text--

What is hoisting in JavaScript?

#### --distractors--

Raising errors to the top of the code

---

The process of creating new variables

---

> Automatically moving 'var' declarations to the top of the block

Reviewer: ```suggestion
Automatically moving `var` declarations to the top of the block
```

#### --answer--

The process of moving variable declarations to the top of their scope

### --question--

#### --text--

What is the purpose of modules in JavaScript?

#### --distractors--

To improve code performance

---

To make code more difficult to understand

---

To create new programming languages

#### --answer--

To organize code into reusable units

### --question--

#### --text--

> In the context of JavaScript modules, what does 'export' do?

Reviewer: ```suggestion
In the context of JavaScript modules, what does `export` do?
```

#### --distractors--

Save the code to disk

---

Include external libraries in your code

---

Publish your code to a public repository

#### --answer--

Make variables, functions or classes available to be imported by other modules

### --question--

#### --text--

Which of these is NOT a valid way to declare a variable in JavaScript?

#### --distractors--

> var myVariable = 'test'

Reviewer: ```suggestion
`var myVariable = 'test'`
```

---

> let myVariable = 'test'

Reviewer: ```suggestion
`let myVariable = 'test'`
```

---

> const myVariable = 'test'

Reviewer: ```suggestion
`const myVariable = 'test'`
```

#### --answer--

> variable myVariable = 'test'

Reviewer: ```suggestion
`variable myVariable = 'test'`
```

### --question--

#### --text--

In the context of JavaScript memory management, what are memory leaks?

#### --distractors--

Oversized memory use

---

Coding errors leading to application crashes

---

Part of the garbage collection process

#### --answer--

> Unreachable pieces of memory by your application but still allocated

Reviewer: ```suggestion
Memory allocations which are never properly made unreachable
```

Something like this would be more accurate.

### --question--

#### --text--

> What is the main difference between 'import' and 'require' in JavaScript?

Reviewer: ```suggestion
What is the main difference between `import` and `require` in JavaScript?
```

#### --distractors--

> 'import' is used in Node.js, while 'require' is used in the browser

Reviewer: ```suggestion
`import` is used in Node.js, while `require` is used in the browser
```

---

> 'import' supports asynchronous loading, while 'require' does not

Reviewer: ```suggestion
`import` supports asynchronous loading, while `require` does not
```

---

> 'import' is used for CSS and image files, while 'require' is used for JavaScript files

Reviewer: ```suggestion
`import` is used for CSS and image files, while `require` is used for JavaScript files
```

#### --answer--

> 'import' is part of ES6 syntax, while 'require' is an older module loading syntax

Reviewer: ```suggestion
`import` is part of ES6 syntax, while `require` is an older module loading syntax
```

### --question--

#### --text--

> What does 'let' provide that 'var' does not?

Reviewer: ```suggestion
What does `let` provide that `var` does not?
```
